### PR TITLE
basic interrupt support, 32 CUs, harden interrrupt in dts

### PR DESCRIPTION
**Pull request by mamin506**
1. Add one interrupt controller to support up to 32 CUs interrupt
2. Harden 32 CUs interrupt for now. Show fix this after we could dynamically get interrupts from XCLBIN.
3. Basic interrupt support, including handler, user configure polling

### CODE REVIEW

**Comment by sonals** [comment]
ok to test

File: src/runtime_src/driver/zynq/drm/zocl/sched_exec.c [hunk 1]
```diff
@@ -1621,6 +1628,41 @@ cq_check(void *data)
...
+		if (zdev->irq[cu_idx] == irq)
+			break;
+	}
+
+	SCHED_DEBUG("cu_idx %d interrupt handle", cu_idx);
+	if (cu_idx >= zdev->cu_num) {
+		/* This should never happen */
+		DRM_ERROR("Unknown isr irq %d, polling %d\n",
+			  irq, zdev->exec->polling_mode);
+		return IRQ_HANDLED;
```
Comment (larry9523) [inline src/runtime_src/driver/zynq/drm/zocl/sched_exec.c]: Seems IRQ_NONE is more proper?

File: src/runtime_src/driver/zynq/drm/zocl/sched_exec.c [hunk 2]
```diff
@@ -1621,6 +1628,41 @@ cq_check(void *data)
...
+	}
+
+	virt_addr = zdev->regs + cu_idx_to_offset(zdev->ddev, cu_idx);
+	/* Clear all CU interrupts, will need to handle ap_done and ap_ready
+	 * interrupt after support dataflow exectution model
+	 */
+	virt_addr[3] = virt_addr[3];
+
+	/* wake up all scheduler ... currently one only */
+	g_sched0.intc = 1;
```
Comment (larry9523) [inline src/runtime_src/driver/zynq/drm/zocl/sched_exec.c]: Do we have race between kds thread setting sched? (sched->ince = 0)

File: src/runtime_src/driver/zynq/drm/zocl/sched_exec.c [hunk 3]
```diff
@@ -1621,6 +1628,41 @@ cq_check(void *data)
...
+		DRM_ERROR("Unknown isr irq %d, polling %d\n",
+			  irq, zdev->exec->polling_mode);
+		return IRQ_HANDLED;
+	}
+
+	virt_addr = zdev->regs + cu_idx_to_offset(zdev->ddev, cu_idx);
+	/* Clear all CU interrupts, will need to handle ap_done and ap_ready
+	 * interrupt after support dataflow exectution model
+	 */
+	virt_addr[3] = virt_addr[3];
```
Comment (larry9523) [inline src/runtime_src/driver/zynq/drm/zocl/sched_exec.c]: This looks weird from software point of view. We'd better add some comments on why we are doing this.

**Comment by larry9523** [comment]
Overall, we only support 32 CU interrupts, We need to add a check, if 33th CU interrupt is configured with interrupt mode, we either return fail or fall back to polling mode.

File: src/runtime_src/driver/zynq/drm/zocl/sched_exec.c (hunk 1 above)
Comment (mamin506) [reply src/runtime_src/driver/zynq/drm/zocl/sched_exec.c]: Done. Thanks

File: src/runtime_src/driver/zynq/drm/zocl/sched_exec.c (hunk 2 above)
Comment (mamin506) [reply src/runtime_src/driver/zynq/drm/zocl/sched_exec.c]: Add comments about the race. In different situation, this should not impact the functionality.

File: src/runtime_src/driver/zynq/drm/zocl/sched_exec.c (hunk 3 above)
Comment (mamin506) [reply src/runtime_src/driver/zynq/drm/zocl/sched_exec.c]: Add comments

**Pull request merged**